src/Spice86.Core/Emulator/OperatingSystem/DosFileManager.cs

```diff
@@ -602,119 +557,56 @@ public class DosFileManager {
     }
 
-    private void SetOpenFile(ushort fileHandle, OpenFile? openFile) {
-        _openFiles[fileHandle] = openFile;
+    private void SetOpenFile(ushort fileHandle, OpenFile? openFile) => OpenFiles[fileHandle] = openFile;
+    
+    private void UpdateDosTransferAreaFromFile(string matchingFile) {
+        if (_loggerService.IsEnabled(Serilog.Events.LogEventLevel.Verbose)) {
+            _loggerService.Verbose("Found matching file {MatchingFile}", matchingFile);
+        }
+        DosDiskTransferArea dosDiskTransferArea = new(_memory, GetDiskTransferAreaAddressPhysical());
+        FileInfo attributes = new FileInfo(matchingFile);
+        DateTime creationZonedDateTime = attributes.CreationTimeUtc;
+        DateTime creationLocalDate = creationZonedDateTime.ToLocalTime();
+        DateTime creationLocalTime = creationZonedDateTime.ToLocalTime();
+        dosDiskTransferArea.FileDate = ToDosDate(creationLocalDate);
+        dosDiskTransferArea.FileTime = ToDosTime(creationLocalTime);
+        dosDiskTransferArea.FileSize = (ushort)attributes.Length;
+        dosDiskTransferArea.FileName = Path.GetFileName(matchingFile);
     }
 
-    private string? ToCaseSensitiveFileName(string? caseInsensitivePath) {
-        if (string.IsNullOrWhiteSpace(caseInsensitivePath)) {
-            return null;
+    /// <summary>
+    /// Creates a directory on disk.
+    /// </summary>
+    /// <param name="dosDirectory">The directory name to create</param>
+    /// <returns></returns>
+    /// <returns>A <see cref="DosFileOperationResult"/> with details about the result of the operation.</returns>
+    public DosFileOperationResult CreateDirectory(string dosDirectory) {
+         string? fullPath = _dosPathResolver.ToHostCaseSensitiveFullName(dosDirectory, true);
+        if (string.IsNullOrWhiteSpace(fullPath)) {
+            return FileNotFoundError(dosDirectory);
         }
 
-        string fileToProcess = ConvertUtils.ToSlashPath(caseInsensitivePath);
-        string? parentDir = Path.GetDirectoryName(fileToProcess);
-        if (File.Exists(fileToProcess) || Directory.Exists(fileToProcess) ||
-            (string.IsNullOrWhiteSpace(parentDir) == false && Directory.Exists(parentDir) && Directory.GetDirectories(parentDir).Length == 0)) {
-            // file exists or root reached, no need to go further. Path found.
-            return caseInsensitivePath;
+        string parentFolder = _dosPathResolver.GetFullNameForParentDirectory(fullPath);
+
+        if (!Directory.Exists(parentFolder)) {
+            CreateDirectory(_dosPathResolver.GetHostRelativePathToCurrentDirectory(parentFolder));
         }
 
-        string? parent = ToCaseSensitiveFileName(parentDir);
-        if (parent == null) {
-            // End of recursion, root reached
-            return null;
+        if (_dosPathResolver.IsThereAnyDirectoryOrFileWithTheSameName(dosDirectory, new DirectoryInfo(parentFolder))) {
+            return FileNotFoundError(dosDirectory);
```

Comment:
I suspect this is not the correct error.

Reply:
This is also susceptible to race conditions, where the file or directory is created between this check and the directory creation.

Reply:
TOCTOU problem, yes. But the host file system is not in our control, isn't it ?

Reply:
It is not. And I think the check is only to see if another file, with a different case, on a case sensitive filesystem, exists. To get an error that the CreateDirectory would not.
So the check is good. It's just the type of error I'm not sure about. Does DOS really throw a FileNotFound if a directory already exists?

Reply:
It is the correct error code. In case of error, it either returns 0x2 (file not found) or 0x5 (access denied, in case of network access, starting with MS-DOS 3.3)

See:
https://www.gladir.com/LEXIQUE/INTR/int21f39.htm
also:
https://marc.info/?l=wine-patches&m=104897290211098&w=2 (msdos/int21.c in wine)
https://gitlab.winehq.org/szanni/wine/-/blob/wine-980517/msdos/int21.c